File: dotnet/src/FluentCards/TeamsAdaptiveCards.cs
namespace FluentCards;

/// <summary>
/// Provides helper methods for creating Microsoft Teams-style Adaptive Cards using the FluentCards fluent API,
/// reflecting common patterns from the Teams Adaptive Card Samples collection.
/// </summary>
public static class TeamsAdaptiveCards
{
    /// <summary>
    /// Creates an approval request card with Approve and Decline actions.
    /// Reflects the Teams approval sample pattern where a requester asks for sign-off on an item.
    /// </summary>
    /// <param name="input">The structured input parameters for the approval card.</param>
    /// <returns>An Adaptive Card representing an approval request.</returns>
    public static AdaptiveCard CreateApprovalCard(ApprovalCardInput input)
    {
        return AdaptiveCardBuilder.Create()
            .WithVersion("1.5")
            .AddColumnSet(cs => cs
                .AddColumn("auto", col => col
                    .AddImage(img => img
                        .WithUrl(input.RequesterImageUrl ?? string.Empty)
                        .WithSize(ImageSize.Small)
                        .WithStyle(ImageStyle.Person)))
                .AddColumn("stretch", col => col
                    .WithVerticalContentAlignment(VerticalAlignment.Center)
                    .AddTextBlock(tb => tb
                        .WithText(input.RequesterName)
                        .WithWeight(TextWeight.Bolder)
                        .WithWrap(true))
                    .AddTextBlock(tb => tb
                        .WithText(input.SubmittedDate)
                        .WithIsSubtle()
                        .WithSize(TextSize.Small)
                        .WithWrap(true))))
            .AddTextBlock(tb => tb
                .WithText(input.Title)
                .WithSize(TextSize.Large)
                .WithWeight(TextWeight.Bolder)
                .WithWrap(true))
            .AddFactSet(fs => fs
                .AddFact("Category", input.Category)
                .AddFact("Amount", input.Amount)
                .AddFact("Business Unit", input.BusinessUnit)
                .AddFact("Due Date", input.DueDate))
            .AddTextBlock(tb => tb
                .WithText(input.Description)
                .WithWrap(true)
                .WithIsSubtle())
            .AddAction(a => a
                .Submit("Approve")
                .WithStyle(ActionStyle.Positive))
            .AddAction(a => a
                .Submit("Decline")
                .WithStyle(ActionStyle.Destructive))
            .Build();
    }

    /// <summary>
    /// Creates a status update notification card showing the current state of a project or task.
    /// Reflects the Teams status-update sample pattern used in project tracking scenarios.
    /// </summary>
    /// <param name="input">The structured input parameters for the status update card.</param>
    /// <returns>An Adaptive Card representing a status update notification.</returns>
    public static AdaptiveCard CreateStatusUpdateCard(StatusUpdateCardInput input)
    {
        return AdaptiveCardBuilder.Create()
            .WithVersion("1.5")
            .AddContainer(c => c
                .WithStyle(ContainerStyle.Emphasis)
                .AddColumnSet(cs => cs
                    .AddColumn("stretch", col => col
                        .AddTextBlock(tb => tb
                            .WithText(input.CardTitle)
                            .WithSize(TextSize.Large)
                            .WithWeight(TextWeight.Bolder)
                            .WithWrap(true))
                        .AddTextBlock(tb => tb
                            .WithText($"{input.TeamName} • {input.UpdateDate}")
                            .WithIsSubtle()
                            .WithSize(TextSize.Small)
                            .WithWrap(true)))))
            .AddFactSet(fs => fs
                .AddFact("Project", input.Project)
                .AddFact("Status", input.Status)
                .AddFact("Sprint", input.Sprint)
                .AddFact("Completion", input.Completion)
                .AddFact("Updated By", input.UpdatedBy))
            .AddTextBlock(tb => tb
                .WithText(input.Notes)
                .WithWrap(true))
            .AddAction(a => a
                .OpenUrl(input.ProjectUrl)
                .WithTitle("View Project"))
            .Build();
    }

    /// <summary>
    /// Creates a task assignment notification card informing the recipient of a newly assigned task.
    /// Reflects the Teams task-update sample pattern used in work tracking integrations.
    /// </summary>
    /// <param name="input">The structured input parameters for the task update card.</param>
    /// <returns>An Adaptive Card representing a task assignment notification.</returns>
    public static AdaptiveCard CreateTaskUpdateCard(TaskUpdateCardInput input)
    {
        return AdaptiveCardBuilder.Create()
            .WithVersion("1.5")
            .AddColumnSet(cs => cs
                .AddColumn("stretch", col => col
                    .AddTextBlock(tb => tb
                        .WithText("Task Assigned to You")
                        .WithSize(TextSize.Large)
                        .WithWeight(TextWeight.Bolder)
                        .WithWrap(true)))
                .AddColumn("auto", col => col
                    .WithVerticalContentAlignment(VerticalAlignment.Center)
                    .AddTextBlock(tb => tb
                        .WithText(input.Priority)
                        .WithColor(TextColor.Attention)
                        .WithWeight(TextWeight.Bolder))))
            .AddFactSet(fs => fs
                .AddFact("Task", input.TaskName)
                .AddFact("Project", input.Project)
                .AddFact("Assigned By", input.AssignedBy)
                .AddFact("Due Date", input.DueDate)
                .AddFact("Estimate", input.Estimate))
            .AddTextBlock(tb => tb
                .WithText(input.Description)
                .WithWrap(true)
                .WithIsSubtle())
            .AddAction(a => a
                .OpenUrl(input.TaskUrl)
                .WithTitle("View Task"))
            .AddAction(a => a
                .Submit("Acknowledge")
                .WithStyle(ActionStyle.Positive))
            .Build();
    }

    /// <summary>
    /// Creates a meeting reminder card with meeting details and a join link.
    /// Reflects the Teams meeting-invite sample pattern used in calendar integration scenarios.
    /// </summary>
    /// <param name="input">The structured input parameters for the meeting reminder card.</param>
    /// <returns>An Adaptive Card representing a meeting reminder.</returns>
    public static AdaptiveCard CreateMeetingReminderCard(MeetingReminderCardInput input)
    {
        return AdaptiveCardBuilder.Create()
            .WithVersion("1.5")
            .AddTextBlock(tb => tb
                .WithText("⏰ Meeting Starting Soon")
                .WithSize(TextSize.Large)
                .WithWeight(TextWeight.Bolder)
                .WithWrap(true))
            .AddTextBlock(tb => tb
                .WithText(input.MeetingTitle)
                .WithSize(TextSize.Medium)
                .WithWrap(true))
            .AddFactSet(fs => fs
                .AddFact("Organizer", input.Organizer)
                .AddFact("Date", input.Date)
                .AddFact("Time", input.Time)
                .AddFact("Location", input.Location)
                .AddFact("Attendees", input.Attendees))
            .AddTextBlock(tb => tb
                .WithText(input.Agenda)
                .WithWrap(true)
                .WithIsSubtle())
            .AddAction(a => a
                .OpenUrl(input.JoinUrl)
                .WithTitle("Join Meeting")
                .WithStyle(ActionStyle.Positive))
            .AddAction(a => a
                .OpenUrl(input.DetailsUrl)
                .WithTitle("View Details"))
            .Build();
    }

    /// <summary>
    /// Creates an expense report card for finance team review with Approve and Reject actions.
    /// Reflects the Teams expense-report sample pattern used in finance approval workflows.
    /// </summary>
    /// <param name="input">The structured input parameters for the expense report card.</param>
    /// <returns>An Adaptive Card representing an expense report for review.</returns>
    public static AdaptiveCard CreateExpenseReportCard(ExpenseReportCardInput input)
    {
        return AdaptiveCardBuilder.Create()
            .WithVersion("1.5")
            .AddContainer(c => c
                .WithStyle(ContainerStyle.Emphasis)
                .AddTextBlock(tb => tb
                    .WithText("Expense Report Submitted")
                    .WithSize(TextSize.Large)
                    .WithWeight(TextWeight.Bolder)
                    .WithWrap(true))
                .AddTextBlock(tb => tb
                    .WithText("Awaiting your review and approval")
                    .WithIsSubtle()
                    .WithWrap(true)))
            .AddColumnSet(cs => cs
                .AddColumn("auto", col => col
                    .AddImage(img => img
                        .WithUrl(input.EmployeeImageUrl ?? string.Empty)
                        .WithSize(ImageSize.Small)
                        .WithStyle(ImageStyle.Person)))
                .AddColumn("stretch", col => col
                    .WithVerticalContentAlignment(VerticalAlignment.Center)
                    .AddTextBlock(tb => tb
                        .WithText(input.EmployeeName)
                        .WithWeight(TextWeight.Bolder)
                        .WithWrap(true))
                    .AddTextBlock(tb => tb
                        .WithText(input.EmployeeJobTitle)
                        .WithIsSubtle()
                        .WithSize(TextSize.Small)
                        .WithWrap(true))))
            .AddFactSet(fs => fs
                .AddFact("Report ID", input.ReportId)
                .AddFact("Submitted", input.SubmittedDate)
                .AddFact("Category", input.Category)
                .AddFact("Total Amount", input.TotalAmount)
                .AddFact("Currency", input.Currency))
            .AddTextBlock(tb => tb
                .WithText(input.Description)
                .WithWrap(true)
                .WithIsSubtle())
            .AddAction(a => a
                .Submit("Approve")
                .WithStyle(ActionStyle.Positive))
            .AddAction(a => a
                .Submit("Reject")
                .WithStyle(ActionStyle.Destructive))
            .AddAction(a => a
                .OpenUrl(input.ReportUrl)
                .WithTitle("View Report"))
            .Build();
    }
}

File: node/fluent-cards/src/TeamsAdaptiveCards.ts
import type { AdaptiveCard } from './models.js';
import { ActionStyle, ContainerStyle, ImageSize, ImageStyle, TextColor, TextSize, TextWeight, VerticalAlignment } from './enums.js';
import { AdaptiveCardBuilder } from './builders/AdaptiveCardBuilder.js';
import { ColumnSetBuilder } from './builders/ColumnSetBuilder.js';

// ─── Input types ─────────────────────────────────────────────────────────────

/** Input parameters for {@link TeamsAdaptiveCards.createApprovalCard}. */
export interface ApprovalCardInput {
  requesterName: string;
  submittedDate: string;
  title: string;
  category: string;
  amount: string;
  businessUnit: string;
  dueDate: string;
  description: string;
  requesterImageUrl?: string;
}

/** Input parameters for {@link TeamsAdaptiveCards.createStatusUpdateCard}. */
export interface StatusUpdateCardInput {
  cardTitle: string;
  teamName: string;
  updateDate: string;
  project: string;
  status: string;
  sprint: string;
  completion: string;
  updatedBy: string;
  notes: string;
  projectUrl: string;
}

/** Input parameters for {@link TeamsAdaptiveCards.createTaskUpdateCard}. */
export interface TaskUpdateCardInput {
  taskName: string;
  project: string;
  assignedBy: string;
  dueDate: string;
  estimate: string;
  priority: string;
  description: string;
  taskUrl: string;
}

/** Input parameters for {@link TeamsAdaptiveCards.createMeetingReminderCard}. */
export interface MeetingReminderCardInput {
  meetingTitle: string;
  organizer: string;
  date: string;
  time: string;
  location: string;
  attendees: string;
  agenda: string;
  joinUrl: string;
  detailsUrl: string;
}

/** Input parameters for {@link TeamsAdaptiveCards.createExpenseReportCard}. */
export interface ExpenseReportCardInput {
  employeeName: string;
  employeeJobTitle: string;
  reportId: string;
  submittedDate: string;
  category: string;
  totalAmount: string;
  currency: string;
  description: string;
  reportUrl: string;
  employeeImageUrl?: string;
}

// ─── Factory methods ─────────────────────────────────────────────────────────

/**
 * Provides helper methods for creating Microsoft Teams-style Adaptive Cards,
 * reflecting common patterns from the Teams Adaptive Card Samples collection.
 */
export const TeamsAdaptiveCards = {
  /**
   * Creates an approval request card with Approve and Decline actions.
   * Reflects the Teams approval sample pattern where a requester asks for sign-off on an item.
   */
  createApprovalCard(input: ApprovalCardInput): AdaptiveCard {
    const b = AdaptiveCardBuilder.create().withVersion('1.5');
    b.addColumnSet(cs => {
      if (input.requesterImageUrl) {
        cs.addColumn('auto', col => col
          .addImage(img => img
            .withUrl(input.requesterImageUrl!)
            .withSize(ImageSize.Small)
            .withStyle(ImageStyle.Person)));
      }
      cs.addColumn('stretch', col => col
        .withVerticalContentAlignment(VerticalAlignment.Center)
        .addTextBlock(tb => tb
          .withText(input.requesterName)
          .withWeight(TextWeight.Bolder)
          .withWrap(true))
        .addTextBlock(tb => tb
          .withText(input.submittedDate)
          .withIsSubtle()
          .withSize(TextSize.Small)
          .withWrap(true)));
    });
    return b
      .addTextBlock(tb => tb
        .withText(input.title)
        .withSize(TextSize.Large)
        .withWeight(TextWeight.Bolder)
        .withWrap(true))
      .addFactSet(fs => fs
        .addFact('Category', input.category)
        .addFact('Amount', input.amount)
        .addFact('Business Unit', input.businessUnit)
        .addFact('Due Date', input.dueDate))
      .addTextBlock(tb => tb
        .withText(input.description)
        .withWrap(true)
        .withIsSubtle())
      .addAction(a => a
        .submit('Approve')
        .withStyle(ActionStyle.Positive))
      .addAction(a => a
        .submit('Decline')
        .withStyle(ActionStyle.Destructive))
      .build();
  },

  /**
   * Creates a status update notification card showing the current state of a project or task.
   * Reflects the Teams status-update sample pattern used in project tracking scenarios.
   */
  createStatusUpdateCard(input: StatusUpdateCardInput): AdaptiveCard {
    return AdaptiveCardBuilder.create()
      .withVersion('1.5')
      .addContainer(c => c
        .withStyle(ContainerStyle.Emphasis)
        .addColumnSet(cs => cs
          .addColumn('stretch', col => col
            .addTextBlock(tb => tb
              .withText(input.cardTitle)
              .withSize(TextSize.Large)
              .withWeight(TextWeight.Bolder)
              .withWrap(true))
            .addTextBlock(tb => tb
              .withText(`${input.teamName} • ${input.updateDate}`)
              .withIsSubtle()
              .withSize(TextSize.Small)
              .withWrap(true)))))
      .addFactSet(fs => fs
        .addFact('Project', input.project)
        .addFact('Status', input.status)
        .addFact('Sprint', input.sprint)
        .addFact('Completion', input.completion)
        .addFact('Updated By', input.updatedBy))
      .addTextBlock(tb => tb
        .withText(input.notes)
        .withWrap(true))
      .addAction(a => a
        .openUrl(input.projectUrl)
        .withTitle('View Project'))
      .build();
  },

  /**
   * Creates a task assignment notification card informing the recipient of a newly assigned task.
   * Reflects the Teams task-update sample pattern used in work tracking integrations.
   */
  createTaskUpdateCard(input: TaskUpdateCardInput): AdaptiveCard {
    return AdaptiveCardBuilder.create()
      .withVersion('1.5')
      .addColumnSet(cs => cs
        .addColumn('stretch', col => col
          .addTextBlock(tb => tb
            .withText('Task Assigned to You')
            .withSize(TextSize.Large)
            .withWeight(TextWeight.Bolder)
            .withWrap(true)))
        .addColumn('auto', col => col
          .withVerticalContentAlignment(VerticalAlignment.Center)
          .addTextBlock(tb => tb
            .withText(input.priority)
            .withColor(TextColor.Attention)
            .withWeight(TextWeight.Bolder))))
      .addFactSet(fs => fs
        .addFact('Task', input.taskName)
        .addFact('Project', input.project)
        .addFact('Assigned By', input.assignedBy)
        .addFact('Due Date', input.dueDate)
        .addFact('Estimate', input.estimate))
      .addTextBlock(tb => tb
        .withText(input.description)
        .withWrap(true)
        .withIsSubtle())
      .addAction(a => a
        .openUrl(input.taskUrl)
        .withTitle('View Task'))
      .addAction(a => a
        .submit('Acknowledge')
        .withStyle(ActionStyle.Positive))
      .build();
  },

  /**
   * Creates a meeting reminder card with meeting details and a join link.
   * Reflects the Teams meeting-invite sample pattern used in calendar integration scenarios.
   */
  createMeetingReminderCard(input: MeetingReminderCardInput): AdaptiveCard {
    return AdaptiveCardBuilder.create()
      .withVersion('1.5')
      .addTextBlock(tb => tb
        .withText('⏰ Meeting Starting Soon')
        .withSize(TextSize.Large)
        .withWeight(TextWeight.Bolder)
        .withWrap(true))
      .addTextBlock(tb => tb
        .withText(input.meetingTitle)
        .withSize(TextSize.Medium)
        .withWrap(true))
      .addFactSet(fs => fs
        .addFact('Organizer', input.organizer)
        .addFact('Date', input.date)
        .addFact('Time', input.time)
        .addFact('Location', input.location)
        .addFact('Attendees', input.attendees))
      .addTextBlock(tb => tb
        .withText(input.agenda)
        .withWrap(true)
        .withIsSubtle())
      .addAction(a => a
        .openUrl(input.joinUrl)
        .withTitle('Join Meeting')
        .withStyle(ActionStyle.Positive))
      .addAction(a => a
        .openUrl(input.detailsUrl)
        .withTitle('View Details'))
      .build();
  },

  /**
   * Creates an expense report card for finance team review with Approve and Reject actions.
   * Reflects the Teams expense-report sample pattern used in finance approval workflows.
   */
  createExpenseReportCard(input: ExpenseReportCardInput): AdaptiveCard {
    const b = AdaptiveCardBuilder.create()
      .withVersion('1.5')
      .addContainer(c => c
        .withStyle(ContainerStyle.Emphasis)
        .addTextBlock(tb => tb
          .withText('Expense Report Submitted')
          .withSize(TextSize.Large)
          .withWeight(TextWeight.Bolder)
          .withWrap(true))
        .addTextBlock(tb => tb
          .withText('Awaiting your review and approval')
          .withIsSubtle()
          .withWrap(true)));
    b.addColumnSet(cs => {
      if (input.employeeImageUrl) {
        cs.addColumn('auto', col => col
          .addImage(img => img
            .withUrl(input.employeeImageUrl!)
            .withSize(ImageSize.Small)
            .withStyle(ImageStyle.Person)));
      }
      cs.addColumn('stretch', col => col
        .withVerticalContentAlignment(VerticalAlignment.Center)
        .addTextBlock(tb => tb
          .withText(input.employeeName)
          .withWeight(TextWeight.Bolder)
          .withWrap(true))
        .addTextBlock(tb => tb
          .withText(input.employeeJobTitle)
          .withIsSubtle()
          .withSize(TextSize.Small)
          .withWrap(true)));
    });
    return b
      .addFactSet(fs => fs
        .addFact('Report ID', input.reportId)
        .addFact('Submitted', input.submittedDate)
        .addFact('Category', input.category)
        .addFact('Total Amount', input.totalAmount)
        .addFact('Currency', input.currency))
      .addTextBlock(tb => tb
        .withText(input.description)
        .withWrap(true)
        .withIsSubtle())
      .addAction(a => a
        .submit('Approve')
        .withStyle(ActionStyle.Positive))
      .addAction(a => a
        .submit('Reject')
        .withStyle(ActionStyle.Destructive))
      .addAction(a => a
        .openUrl(input.reportUrl)
        .withTitle('View Report'))
      .build();
  },
};
